File: Backend/pdf-extractor/src/main/java/pdf/extractor/service/PdfService.java
package pdf.extractor.service;

import org.apache.pdfbox.pdmodel.PDDocument;
import org.apache.pdfbox.text.PDFTextStripper;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import org.springframework.http.ResponseEntity;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.web.util.UriComponentsBuilder;
// hola
import java.io.IOException;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Service
public class PdfService {

    private static final String GOOGLE_BOOKS_API_URL = "https://www.googleapis.com/books/v1/volumes";

    public String processPdf(MultipartFile file, String bookTitle) {
        try {
            if (bookTitle != null && !bookTitle.isEmpty()) {
                return searchBookInGoogleBooks(bookTitle);
            }
            String extractedTitle = extractBookNameFromPdf(file);

            if (extractedTitle.isEmpty()) {
                return "No se pudo encontrar un título en el PDF.";
            }

            return searchBookInGoogleBooks(extractedTitle);
        } catch (IOException e) {
            e.printStackTrace();
            return "Error al procesar el PDF.";
        }
    }

    private String extractBookNameFromPdf(MultipartFile file) throws IOException {
        PDDocument document = PDDocument.load(file.getInputStream());
        PDFTextStripper stripper = new PDFTextStripper();

        String text = stripper.getText(document);
        document.close();

        // Normalizar el texto
        text = text.replaceAll("[^a-zA-ZáéíóúÁÉÍÓÚñÑ0-9\\s]", ""); // Eliminar caracteres especiales
        text = text.replaceAll("\\s+", " ").trim(); // Eliminar espacios extra

        // Expresión regular para encontrar títulos en cualquier parte del PDF
        Pattern titlePattern = Pattern
                .compile("(?m)^(?:[A-ZÁÉÍÓÚÑ][a-záéíóúñ]+(?:\\s+[A-ZÁÉÍÓÚÑa-záéíóúñ0-9]+){1,5})$");
        Matcher matcher = titlePattern.matcher(text);

        while (matcher.find()) {
            String possibleTitle = matcher.group().trim();
            if (possibleTitle.length() > 3 && possibleTitle.length() < 100) {
                return possibleTitle; // Devuelve el primer título que encuentre
            }
        }

        // Si no encuentra un título, usa la línea más larga como posible título
        String[] lines = text.split("\\n");
        String longestLine = "";
        for (String line : lines) {
            if (line.length() > longestLine.length()) {
                longestLine = line.trim();
            }
        }

        return longestLine.isEmpty() ? "Título no encontrado" : longestLine;
    }

    private String searchBookInGoogleBooks(String bookName) {
        UriComponentsBuilder uriBuilder = UriComponentsBuilder.fromHttpUrl(GOOGLE_BOOKS_API_URL)
                .queryParam("q", bookName);

        RestTemplate restTemplate = new RestTemplate();
        ResponseEntity<String> response = restTemplate.getForEntity(uriBuilder.toUriString(), String.class);

        if (response.getStatusCode().is2xxSuccessful()) {
            return response.getBody();
        }

        // Si la solicitud falla, intenta con el títolo sin acentos
        String bookNameWithoutAccents = bookName.replaceAll("[áéíóú]", "a");
        uriBuilder = UriComponentsBuilder.fromHttpUrl(GOOGLE_BOOKS_API_URL)
                .queryParam("q", bookNameWithoutAccents);

        response = restTemplate.getForEntity(uriBuilder.toUriString(), String.class);

        return response.getBody();
    }

}

File: services/pdf_service.py
import pdfplumber
import re
import requests
from services.google_books import search_book_in_google_books

async def process_pdf(file, bookTitle=None):
    if bookTitle:
        return search_book_in_google_books(bookTitle)

    extracted_title = await extract_book_name_from_pdf(file)

    if not extracted_title:
        return "No se pudo encontrar un título en el PDF."

    return search_book_in_google_books(extracted_title)

async def extract_book_name_from_pdf(file):
    with pdfplumber.open(file.file) as pdf:
        text = "\n".join(page.extract_text() for page in pdf.pages if page.extract_text())

    # Normalizar el texto
    text = re.sub(r"[^a-zA-ZáéíóúÁÉÍÓÚñÑ0-9\s]", "", text)
    text = re.sub(r"\s+", " ", text).strip()

    # Expresión regular para detectar títulos
    title_pattern = re.compile(r"(?m)^(?:[A-ZÁÉÍÓÚÑ][a-záéíóúñ]+(?:\s+[A-ZÁÉÍÓÚÑa-záéíóúñ0-9]+){1,5})$")
    matches = title_pattern.findall(text)

    for possible_title in matches:
        if 3 < len(possible_title) < 100:
            return possible_title.strip()

    # Si no encuentra título, usa la línea más larga
    lines = text.split("\n")
    longest_line = max(lines, key=len, default="Título no encontrado").strip()

    return longest_line
